test/catalog/catalog_test.cpp

```diff
@@ -502,4 +502,64 @@ TEST_F(CatalogTests, GetIndexObjectsTest) {
   txn_manager_->Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
 }
 
+/*
+ * Exercise a bunch of scenarios of the DDL lock semantics
+ */
+// NOLINTNEXTLINE
+TEST_F(CatalogTests, DDLLockTest) {
+  // txn2 is used to verify that older txns can't acquire the lock
+  auto *txn2 = txn_manager_->BeginTransaction();
+
+  // txn0 is used to verify conflict scenarios and that commits release the lock
+  auto *txn0 = txn_manager_->BeginTransaction();
+  auto accessor0 = catalog_->GetAccessor(txn0, db_);
+  EXPECT_NE(accessor0, nullptr);
+  auto ns_oid = accessor0->CreateNamespace("test_namespace0");  // Should succeed as txn0 acquires the lock
+  EXPECT_NE(ns_oid, catalog::INVALID_NAMESPACE_OID);
+  VerifyCatalogTables(*accessor0);  // Check visibility to me
+
+  ns_oid = accessor0->CreateNamespace("test_namespace4");  // Should work since txn0 already holds lock
+  EXPECT_NE(ns_oid, catalog::INVALID_NAMESPACE_OID);
+  VerifyCatalogTables(*accessor0);  // Check visibility to me
+
+  // txn1 is used to verify that concurrent txns cannot acquire lock, even if a newer timestamp
+  auto *txn1 = txn_manager_->BeginTransaction();
+  auto accessor1 = catalog_->GetAccessor(txn1, db_);
+  ns_oid = accessor1->CreateNamespace("test_namespace1");
+  EXPECT_EQ(ns_oid, catalog::INVALID_NAMESPACE_OID);  // Should fail to get the lock because txn0 holds it (txn1 > txn0)
+  txn_manager_->Abort(txn1);
+
+  auto accessor2 = catalog_->GetAccessor(txn2, db_);
+  ns_oid = accessor2->CreateNamespace("test_namespace2");
+  EXPECT_EQ(ns_oid, catalog::INVALID_NAMESPACE_OID);  // Should fail to get the lock because txn0 holds it (txn2 < txn0)
+
+  txn_manager_->Commit(txn0, transaction::TransactionUtil::EmptyCallback, nullptr);  // txn0 releases the lock
+
+  ns_oid = accessor2->CreateNamespace("test_namespace2");
```

Comment:
Semantic violation of failure makes this undefined behavior.  The first failure should force `txn2` to abort so this should fail regardless of the lock state.